File: backend/app/schemas/rbac.py
"""RBAC (Role-Based Access Control) schemas."""

from datetime import datetime
from pydantic import BaseModel, ConfigDict, Field


# Permission Schemas
class PermissionBase(BaseModel):
    """Base permission schema."""

    name: str = Field(
        ...,
        min_length=1,
        max_length=100,
        description="Permission name (e.g., 'user.create')",
    )
    resource: str = Field(
        ...,
        min_length=1,
        max_length=50,
        description="Resource type (e.g., 'user', 'recommendation')",
    )
    action: str = Field(
        ...,
        min_length=1,
        max_length=50,
        description="Action type (e.g., 'create', 'read', 'update', 'delete')",
    )
    description: str | None = Field(None, description="Permission description")


class PermissionCreate(PermissionBase):
    """Schema for creating a permission."""

    pass


class PermissionUpdate(BaseModel):
    """Schema for updating a permission."""

    name: str | None = Field(None, min_length=1, max_length=100)
    resource: str | None = Field(None, min_length=1, max_length=50)
    action: str | None = Field(None, min_length=1, max_length=50)
    description: str | None = None


class Permission(PermissionBase):
    """Schema for permission response."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    created_at: datetime
    updated_at: datetime


class PermissionListResponse(BaseModel):
    """Schema for paginated permission list response."""

    items: list[Permission]
    total: int
    limit: int
    offset: int


# Role Schemas
class RoleBase(BaseModel):
    """Base role schema."""

    name: str = Field(..., min_length=1, max_length=50, description="Role name")
    description: str | None = Field(None, description="Role description")


class RoleCreate(RoleBase):
    """Schema for creating a role."""

    permission_ids: list[int] = Field(
        default_factory=list, description="List of permission IDs to assign"
    )


class RoleUpdate(BaseModel):
    """Schema for updating a role."""

    name: str | None = Field(None, min_length=1, max_length=50)
    description: str | None = None
    permission_ids: list[int] | None = Field(
        None, description="List of permission IDs to assign (replaces existing)"
    )


class Role(RoleBase):
    """Schema for role response."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    created_at: datetime
    updated_at: datetime
    permissions: list[Permission] = Field(default_factory=list)


class RoleListResponse(BaseModel):
    """Schema for paginated role list response."""

    items: list[Role]
    total: int
    limit: int
    offset: int


# Role-Permission Assignment
class RolePermissionAssignment(BaseModel):
    """Schema for assigning/removing permissions to/from a role."""

    permission_ids: list[int] = Field(
        ..., min_items=1, description="List of permission IDs"
    )

File: frontend/src/types/rbac.ts
/**
 * Role-Based Access Control (RBAC) types
 */

export enum UserRole {
  SUPER_ADMIN = 'super_admin',
  ADMIN = 'admin',
  USER = 'user',
  GUEST = 'guest',
}

export interface User {
  id: number
  email: string
  full_name: string | null
  is_active: boolean
  role: UserRole
  created_at: string
  updated_at: string

  // Profile fields (optional)
  gender?: string | null
  age?: number | null
  identity?: string | null
  style?: string | null
  temperature_sensitivity?: string | null
  activity_context?: string | null
  other_preferences?: string | null
}

export interface UserAdminCreate {
  email: string
  password: string
  full_name?: string | null
  role?: UserRole
  is_active?: boolean
}

export interface UserAdminUpdate {
  full_name?: string | null
  email?: string | null
  role?: UserRole
  is_active?: boolean
  password?: string | null

  // Profile fields (optional)
  gender?: string | null
  age?: number | null
  identity?: string | null
  style?: string | null
  temperature_sensitivity?: string | null
  activity_context?: string | null
  other_preferences?: string | null
}

export interface RoleUpdateRequest {
  role: UserRole
}

export interface UserListResponse {
  items: User[]
  total: number
  limit: number
  offset: number
}

export interface UserListParams {
  skip?: number
  limit?: number
  search?: string
  role?: UserRole
}

// Permission types for RBAC
export interface Permission {
  id: number
  name: string
  resource: string
  action: string
  description: string | null
  created_at: string
  updated_at: string
}

export interface PermissionCreate {
  name: string
  resource: string
  action: string
  description?: string | null
}

export interface PermissionUpdate {
  name?: string
  resource?: string
  action?: string
  description?: string | null
}

export interface PermissionListResponse {
  items: Permission[]
  total: number
  limit: number
  offset: number
}

export interface PermissionListParams {
  skip?: number
  limit?: number
  resource?: string
}

// Role types for RBAC
export interface Role {
  id: number
  name: string
  description: string | null
  created_at: string
  updated_at: string
  permissions: Permission[]
}

export interface RoleCreate {
  name: string
  description?: string | null
  permission_ids?: number[]
}

export interface RoleUpdate {
  name?: string
  description?: string | null
  permission_ids?: number[]
}

export interface RoleListResponse {
  items: Role[]
  total: number
  limit: number
  offset: number
}

export interface RoleListParams {
  skip?: number
  limit?: number
  search?: string
}

export interface RolePermissionAssignment {
  permission_ids: number[]
}
